File: source code/backend/ai-powered-software-testing/Chapter 10/Models/Responses/ResponseModels.cs

using Chapter_10.Models.Requests;

namespace Chapter_10.Models.Responses
{
    public class MetricDesignResponse
    {
        public string DesignId { get; set; }
        public DesignedMetric[] Metrics { get; set; }
        public MetricRelationship[] Relationships { get; set; }
        public CollectionPlan CollectionPlan { get; set; }
        public InterpretationFramework InterpretationFramework { get; set; }
        public ValidationRule[] ValidationRules { get; set; }
        public ImplementationGuidance ImplementationGuidance { get; set; }
        public SuccessCriteria SuccessCriteria { get; set; }
    }

    public class DesignedMetric
    {
        public string MetricId { get; set; }
        public string Name { get; set; }
        public string Description { get; set; }
        public string Formula { get; set; }
        public string DataSource { get; set; }
        public string[] BusinessObjectives { get; set; }
        public double TargetValue { get; set; }
        public string Unit { get; set; }
        public string[] Dimensions { get; set; }
        public double Weight { get; set; }
        public string Category { get; set; }
    }

    public class MetricRelationship
    {
        public string SourceMetricId { get; set; }
        public string TargetMetricId { get; set; }
        public string RelationshipType { get; set; } // "correlation", "causation", "composition"
        public double Strength { get; set; }
    }

    public class CollectionPlan
    {
        public string[] CollectionSteps { get; set; }
        public string[] Tools { get; set; }
        public string Frequency { get; set; }
        public string ResponsibleTeam { get; set; }
    }

    public class InterpretationFramework
    {
        public string[] InterpretationRules { get; set; }
        public ThresholdDefinition[] Thresholds { get; set; }
        public string[] OutlierHandling { get; set; }
    }

    public class ThresholdDefinition
    {
        public string MetricId { get; set; }
        public double Green { get; set; }
        public double Yellow { get; set; }
        public double Red { get; set; }
    }

    public class ValidationRule
    {
        public string MetricId { get; set; }
        public string Rule { get; set; }
        public string ErrorMessage { get; set; }
    }

    public class ImplementationGuidance
    {
        public string[] Prerequisites { get; set; }
        public string[] Steps { get; set; }
        public string[] Pitfalls { get; set; }
        public string[] SuccessFactors { get; set; }
    }

    public class SuccessCriteria
    {
        public string[] QuantitativeCriteria { get; set; }
        public string[] QualitativeCriteria { get; set; }
        public string ReviewPeriod { get; set; }
    }
}

// Models/Responses/HealthScoreResponse.cs
namespace Chapter_10.Models.Responses
{
    public class HealthScoreResponse
    {
        public string HealthScoreId { get; set; }
        public double OverallScore { get; set; }
        public ComponentScore[] ComponentScores { get; set; }
        public double Confidence { get; set; }
        public ContributingFactor[] ContributingFactors { get; set; }
        public ImprovementRecommendation[] Recommendations { get; set; }
        public TrendAnalysis TrendAnalysis { get; set; }
        public BenchmarkComparison BenchmarkComparison { get; set; }
        public AlertStatus AlertStatus { get; set; }
    }

    public class ComponentScore
    {
        public string ComponentName { get; set; }
        public double Score { get; set; }
        public double Weight { get; set; }
        public string Status { get; set; }
    }

    public class ContributingFactor
    {
        public string Factor { get; set; }
        public double Impact { get; set; }
        public string Direction { get; set; } // "positive", "negative"
    }

    public class ImprovementRecommendation
    {
        public string MetricId { get; set; }
        public string Recommendation { get; set; }
        public double Impact { get; set; }
        public string Priority { get; set; }
        public string[] Steps { get; set; }
    }

    public class TrendAnalysis
    {
        public string Direction { get; set; }
        public double Rate { get; set; }
        public string[] Observations { get; set; }
    }

    public class BenchmarkComparison
    {
        public double Percentile { get; set; }
        public double GapToAverage { get; set; }
        public double GapToBest { get; set; }
    }

    public class AlertStatus
    {
        public string Level { get; set; }
        public string[] Alerts { get; set; }
        public string RecommendedAction { get; set; }
    }
}

// Models/Responses/PredictionResponse.cs
namespace Chapter_10.Models.Responses
{
    public class PredictionResponse
    {
        public string PredictionId { get; set; }
        public CurrentStateAnalysis CurrentState { get; set; }
        public MetricPrediction[] Predictions { get; set; }
        public PredictionConfidence PredictionConfidence { get; set; }
        public Intervention[] Interventions { get; set; }
        public RiskAssessment RiskAssessment { get; set; }
        public MonitoringRecommendation[] MonitoringRecommendations { get; set; }
        public DecisionSupport DecisionSupport { get; set; }
    }

    public class CurrentStateAnalysis
    {
        public DateTime AnalysisDate { get; set; }
        public Dictionary<string, double> CurrentValues { get; set; }
        public string[] Strengths { get; set; }
        public string[] Weaknesses { get; set; }
    }

    public class MetricPrediction
    {
        public string MetricId { get; set; }
        public string MetricName { get; set; }
        public DateTime[] Dates { get; set; }
        public double[] Values { get; set; }
        public double[] LowerBound { get; set; }
        public double[] UpperBound { get; set; }
        public string Trend { get; set; }
    }

    public class PredictionConfidence
    {
        public double AverageConfidence { get; set; }
        public Dictionary<string, double> MetricConfidence { get; set; }
        public string[] ConfidenceFactors { get; set; }
    }

    public class Intervention
    {
        public string MetricId { get; set; }
        public string Type { get; set; }
        public string Description { get; set; }
        public DateTime RecommendedDate { get; set; }
        public double ExpectedImpact { get; set; }
    }

    public class RiskAssessment
    {
        public string OverallRisk { get; set; }
        public string[] IdentifiedRisks { get; set; }
        public string[] MitigationStrategies { get; set; }
    }

    public class MonitoringRecommendation
    {
        public string MetricId { get; set; }
        public string Frequency { get; set; }
        public string Threshold { get; set; }
    }

    public class DecisionSupport
    {
        public string[] DecisionPoints { get; set; }
        public string[] RecommendedActions { get; set; }
        public Dictionary<string, string> Tradeoffs { get; set; }
    }
}

// Models/Responses/InsightResponse.cs
namespace Chapter_10.Models.Responses
{
    public class InsightResponse
    {
        public string InsightId { get; set; }
        public ActionableInsight[] GeneratedInsights { get; set; }
        public InsightImplementationGuidance ImplementationGuidance { get; set; }
        public InsightQuality InsightQuality { get; set; }
        public ValidationStep[] ValidationSteps { get; set; }
        public CommunicationPlan CommunicationPlan { get; set; }
        public InsightMonitoringPlan MonitoringPlan { get; set; }
    }

    public class ActionableInsight
    {
        public string InsightId { get; set; }
        public string Title { get; set; }
        public string Description { get; set; }
        public string Type { get; set; }
        public double ActionabilityScore { get; set; }
        public double Impact { get; set; }
        public string[] AffectedMetrics { get; set; }
        public string[] RecommendedActions { get; set; }
        public string Priority { get; set; }
    }

    public class InsightImplementationGuidance
    {
        public string[] Prerequisites { get; set; }
        public string[] Steps { get; set; }
        public string[] Resources { get; set; }
    }

    public class InsightQuality
    {
        public double Accuracy { get; set; }
        public double Relevance { get; set; }
        public double Timeliness { get; set; }
        public string[] Limitations { get; set; }
    }

    public class ValidationStep
    {
        public string Step { get; set; }
        public string Method { get; set; }
        public string ExpectedOutcome { get; set; }
    }

    public class CommunicationPlan
    {
        public string[] Stakeholders { get; set; }
        public string Message { get; set; }
        public string Channel { get; set; }
        public DateTime Timeline { get; set; }
    }

    public class InsightMonitoringPlan
    {
        public string[] Metrics { get; set; }
        public string ReviewFrequency { get; set; }
        public string SuccessCriteria { get; set; }
    }
}

// Models/Responses/OptimizationResponse.cs
namespace Chapter_10.Models.Responses
{
    public class OptimizationResponse
    {
        public string OptimizationId { get; set; }
        public MetricDefinition[] CurrentMetrics { get; set; }
        public OptimizationRecommendation Optimization { get; set; }
        public ExpectedBenefits ExpectedBenefits { get; set; }
        public ImplementationPlan ImplementationPlan { get; set; }
        public PreservationValidation PreservationValidation { get; set; }
        public OptimizationRiskAssessment RiskAssessment { get; set; }
        public OptimizationMonitoringPlan MonitoringPlan { get; set; }
        public ContinuousImprovementPlan ContinuousImprovement { get; set; }
    }

    public class OptimizationRecommendation
    {
        public RecommendedAction[] RecommendedActions { get; set; }
        public MetricConsolidation[] Consolidations { get; set; }
        public string[] DeprecatedMetrics { get; set; }
        public string[] NewMetrics { get; set; }
    }

    public class RecommendedAction
    {
        public string MetricId { get; set; }
        public string Action { get; set; } // "keep", "modify", "remove", "add"
        public string Rationale { get; set; }
        public double CostSaving { get; set; }
    }

    public class MetricConsolidation
    {
        public string[] SourceMetricIds { get; set; }
        public string TargetMetricId { get; set; }
        public string ConsolidationMethod { get; set; }
    }

    public class ExpectedBenefits
    {
        public double CostReduction { get; set; }
        public double EfficiencyGain { get; set; }
        public double QualityImprovement { get; set; }
        public string[] AdditionalBenefits { get; set; }
    }

    public class ImplementationPlan
    {
        public string[] Phases { get; set; }
        public string Timeline { get; set; }
        public string[] Dependencies { get; set; }
        public string[] RollbackSteps { get; set; }
    }

    public class PreservationValidation
    {
        public bool IsPreserved { get; set; }
        public string[] ValidatedRules { get; set; }
        public string[] Warnings { get; set; }
    }

    public class OptimizationRiskAssessment
    {
        public string[] Risks { get; set; }
        public string[] Mitigations { get; set; }
    }

    public class OptimizationMonitoringPlan
    {
        public string[] KeyIndicators { get; set; }
        public string ReviewFrequency { get; set; }
    }

    public class ContinuousImprovementPlan
    {
        public string[] ReviewCycles { get; set; }
        public string[] FeedbackLoops { get; set; }
    }
}

File: source code/fontend/ai-powered-st/src/chapter-10/models/responses/response-models.ts
import { MetricDefinition } from "../requests/request-models";

// models/responses/metric-design-response.ts
export interface MetricDesignResponse {
    designId: string;
    metrics: DesignedMetric[];
    relationships: MetricRelationship[];
    collectionPlan: CollectionPlan;
    interpretationFramework: InterpretationFramework;
    validationRules: ValidationRule[];
    implementationGuidance: ImplementationGuidance;
    successCriteria: SuccessCriteria;
}

export interface DesignedMetric {
    metricId: string;
    name: string;
    description: string;
    formula: string;
    dataSource: string;
    businessObjectives: string[];
    targetValue: number;
    unit: string;
    dimensions: string[];
    weight: number;
    category: string;
}

export interface MetricRelationship {
    sourceMetricId: string;
    targetMetricId: string;
    relationshipType: string; // "correlation", "causation", "composition"
    strength: number;
}

export interface CollectionPlan {
    collectionSteps: string[];
    tools: string[];
    frequency: string;
    responsibleTeam: string;
}

export interface InterpretationFramework {
    interpretationRules: string[];
    thresholds: ThresholdDefinition[];
    outlierHandling: string[];
}

export interface ThresholdDefinition {
    metricId: string;
    green: number;
    yellow: number;
    red: number;
}

export interface ValidationRule {
    metricId: string;
    rule: string;
    errorMessage: string;
}

export interface ImplementationGuidance {
    prerequisites: string[];
    steps: string[];
    pitfalls: string[];
    successFactors: string[];
}

export interface SuccessCriteria {
    quantitativeCriteria: string[];
    qualitativeCriteria: string[];
    reviewPeriod: string;
}

// models/responses/health-score-response.ts
export interface HealthScoreResponse {
    healthScoreId: string;
    overallScore: number;
    componentScores: ComponentScore[];
    confidence: number;
    contributingFactors: ContributingFactor[];
    recommendations: ImprovementRecommendation[];
    trendAnalysis: TrendAnalysis;
    benchmarkComparison: BenchmarkComparison;
    alertStatus: AlertStatus;
}

export interface ComponentScore {
    componentName: string;
    score: number;
    weight: number;
    status: string;
}

export interface ContributingFactor {
    factor: string;
    impact: number;
    direction: string; // "positive", "negative"
}

export interface ImprovementRecommendation {
    metricId: string;
    recommendation: string;
    impact: number;
    priority: string;
    steps: string[];
}

export interface TrendAnalysis {
    direction: string;
    rate: number;
    observations: string[];
}

export interface BenchmarkComparison {
    percentile: number;
    gapToAverage: number;
    gapToBest: number;
}

export interface AlertStatus {
    level: string;
    alerts: string[];
    recommendedAction: string;
}

// models/responses/prediction-response.ts
export interface PredictionResponse {
    predictionId: string;
    currentState: CurrentStateAnalysis;
    predictions: MetricPrediction[];
    predictionConfidence: PredictionConfidence;
    interventions: Intervention[];
    riskAssessment: RiskAssessment;
    monitoringRecommendations: MonitoringRecommendation[];
    decisionSupport: DecisionSupport;
}

export interface CurrentStateAnalysis {
    analysisDate: Date;
    currentValues: { [key: string]: number };
    strengths: string[];
    weaknesses: string[];
}

export interface MetricPrediction {
    metricId: string;
    metricName: string;
    dates: Date[];
    values: number[];
    lowerBound: number[];
    upperBound: number[];
    trend: string;
}

export interface PredictionConfidence {
    averageConfidence: number;
    metricConfidence: { [key: string]: number };
    confidenceFactors: string[];
}

export interface Intervention {
    metricId: string;
    type: string;
    description: string;
    recommendedDate: Date;
    expectedImpact: number;
}

export interface RiskAssessment {
    overallRisk: string;
    identifiedRisks: string[];
    mitigationStrategies: string[];
}

export interface MonitoringRecommendation {
    metricId: string;
    frequency: string;
    threshold: string;
}

export interface DecisionSupport {
    decisionPoints: string[];
    recommendedActions: string[];
    tradeoffs: { [key: string]: string };
}

// models/responses/insight-response.ts
export interface InsightResponse {
    insightId: string;
    generatedInsights: ActionableInsight[];
    implementationGuidance: InsightImplementationGuidance;
    insightQuality: InsightQuality;
    validationSteps: ValidationStep[];
    communicationPlan: CommunicationPlan;
    monitoringPlan: InsightMonitoringPlan;
}

export interface ActionableInsight {
    insightId: string;
    title: string;
    description: string;
    type: string;
    actionabilityScore: number;
    impact: number;
    affectedMetrics: string[];
    recommendedActions: string[];
    priority: string;
}

export interface InsightImplementationGuidance {
    prerequisites: string[];
    steps: string[];
    resources: string[];
}

export interface InsightQuality {
    accuracy: number;
    relevance: number;
    timeliness: number;
    limitations: string[];
}

export interface ValidationStep {
    step: string;
    method: string;
    expectedOutcome: string;
}

export interface CommunicationPlan {
    stakeholders: string[];
    message: string;
    channel: string;
    timeline: Date;
}

export interface InsightMonitoringPlan {
    metrics: string[];
    reviewFrequency: string;
    successCriteria: string;
}

// models/responses/optimization-response.ts
export interface OptimizationResponse {
    optimizationId: string;
    currentMetrics: MetricDefinition[];
    optimization: OptimizationRecommendation;
    expectedBenefits: ExpectedBenefits;
    implementationPlan: ImplementationPlan;
    preservationValidation: PreservationValidation;
    riskAssessment: OptimizationRiskAssessment;
    monitoringPlan: OptimizationMonitoringPlan;
    continuousImprovement: ContinuousImprovementPlan;
}

export interface OptimizationRecommendation {
    recommendedActions: RecommendedAction[];
    consolidations: MetricConsolidation[];
    deprecatedMetrics: string[];
    newMetrics: string[];
}

export interface RecommendedAction {
    metricId: string;
    action: string; // "keep", "modify", "remove", "add"
    rationale: string;
    costSaving: number;
}

export interface MetricConsolidation {
    sourceMetricIds: string[];
    targetMetricId: string;
    consolidationMethod: string;
}

export interface ExpectedBenefits {
    costReduction: number;
    efficiencyGain: number;
    qualityImprovement: number;
    additionalBenefits: string[];
}

export interface ImplementationPlan {
    phases: string[];
    timeline: string;
    dependencies: string[];
    rollbackSteps: string[];
}

export interface PreservationValidation {
    isPreserved: boolean;
    validatedRules: string[];
    warnings: string[];
}

export interface OptimizationRiskAssessment {
    risks: string[];
    mitigations: string[];
}

export interface OptimizationMonitoringPlan {
    keyIndicators: string[];
    reviewFrequency: string;
}

export interface ContinuousImprovementPlan {
    reviewCycles: string[];
    feedbackLoops: string[];
}

// models/responses/error-response.ts
export interface MetricErrorResponse {
    errorType: string;
    message: string;
    recoverySteps: string[];
    fallbackSuggestion: string;
    diagnosticData?: MetricDiagnosticData;
}

export interface MetricDiagnosticData {
    ambiguousObjectives?: string[];
    clarificationQuestions?: string[];
    inconsistencyDetails?: InconsistencyDetails;
    dataQualityIssues?: DataQualityIssue[];
    patternDetectionChallenges?: string[];
    dataRequirements?: string[];
    insightGenerationChallenges?: string[];
    metricLimitations?: string[];
    conflictingGoals?: string[];
    tradeOffAnalysis?: TradeOffAnalysis;
}

export interface InconsistencyDetails {
    inconsistentMetrics?: string[];
    timePeriodIssues?: string[];
    unitMismatches?: string[];
}

export interface DataQualityIssue {
    metricId: string;
    issue: string;
    severity: string;
}

export interface TradeOffAnalysis {
    tradeoffs?: string[];
    impacts?: { [key: string]: number };
}